Sources/Data/models/WalletUserAsset.swift

```diff
@@ -0,0 +1,181 @@
+  @NSManaged public var isERC20: Bool
+  @NSManaged public var isERC721: Bool
+  @NSManaged public var isERC1155: Bool
+  @NSManaged public var isNFT: Bool
+  @NSManaged public var symbol: String
+  @NSManaged public var decimals: Int32
+  @NSManaged public var visible: Bool
+  @NSManaged public var tokenId: String
+  @NSManaged public var coingeckoId: String
+  @NSManaged public var chainId: String
+  @NSManaged public var coin: Int16
+  @NSManaged public var walletUserAssetGroup: WalletUserAssetGroup?
+  
+  public var blockchainToken: BraveWallet.BlockchainToken {
+    .init(
+      contractAddress: self.contractAddress,
+      name: self.name,
+      logo: self.logo,
+      isErc20: self.isERC20,
+      isErc721: self.isERC721,
+      isErc1155: self.isERC1155,
+      isNft: self.isNFT,
+      symbol: self.symbol,
+      decimals: self.decimals,
+      visible: self.visible,
+      tokenId: self.tokenId,
+      coingeckoId: self.coingeckoId,
+      chainId: self.chainId,
+      coin: BraveWallet.CoinType(rawValue: Int(self.coin))!
+    )
+  }
+  
+  @available(*, unavailable)
+  public init() {
+    fatalError("No Such Initializer: init()")
+  }
+  
+  @available(*, unavailable)
+  public init(context: NSManagedObjectContext) {
+    fatalError("No Such Initializer: init(context:)")
+  }
+  
+  @objc
+  private override init(entity: NSEntityDescription, insertInto context: NSManagedObjectContext?) {
+    super.init(entity: entity, insertInto: context)
+  }
+  
+  public init(context: NSManagedObjectContext, asset: BraveWallet.BlockchainToken) {
+    let entity = Self.entity(context)
+    super.init(entity: entity, insertInto: context)
+    self.contractAddress = asset.contractAddress
+    self.name = asset.name
+    self.logo = asset.logo
+    self.isERC20 = asset.isErc20
+    self.isERC721 = asset.isErc721
+    self.isERC1155 = asset.isErc1155
+    self.isNFT = asset.isNft
+    self.symbol = asset.symbol
+    self.decimals = asset.decimals
+    self.visible = asset.visible
+    self.tokenId = asset.tokenId
+    self.coingeckoId = asset.coingeckoId
+    self.chainId = asset.chainId
+    self.coin = Int16(asset.coin.rawValue)
+  }
+  
+  public static func getUserAsset(asset: BraveWallet.BlockchainToken, context: NSManagedObjectContext? = nil) -> WalletUserAsset? {
+    WalletUserAsset.first(where: NSPredicate(format: "contractAddress == %@ AND chainId == %@ AND symbol == %@ AND tokenId == %@", asset.contractAddress, asset.chainId, asset.symbol, asset.tokenId), context: context ?? DataController.viewContext)
+  }
+  
+  public static func getAllVisibleUserAssets(context: NSManagedObjectContext? = nil) -> [WalletUserAsset]? {
+    WalletUserAsset.all(where: NSPredicate(format: "visible = true"), context: context ?? DataController.viewContext)
+  }
+  
+  public static func migrateVisibleAssets(_ assets: [String: [BraveWallet.BlockchainToken]], completion: (() -> Void)? = nil) {
+    for groupId in assets.keys {
```

Comment:
Is this correct?
You want to "DataController.perform" multiple times in this loop, and call "completion" multiple times?

Otherwise we should swap these.
Do DataController.perform. Then inside it, you do your loop. After the loop, you save once, and then call completion :) 

Result will be much better performance too.

Reply:
updated in [5c0870b](https://github.com/brave/brave-ios/pull/7534/commits/5c0870b6b40e4664558ca6c5764f861aa1e01b4d)